src/vm/methodtable.cpp

```diff
@@ -2343,6 +2342,387 @@ SystemVClassificationType MethodTable::ReClassifyField(SystemVClassificationType
+    {
+        sortedFieldOrder[i] = -1;
+    }
+
+    for (unsigned i = 0; i < helperPtr->currentUniqueOffsetField; i++)
+    {
+        _ASSERTE(helperPtr->fieldOffsets[i] < SYSTEMV_MAX_NUM_FIELDS_IN_REGISTER_PASSED_STRUCT);
+        _ASSERTE(sortedFieldOrder[helperPtr->fieldOffsets[i]] == -1); // we haven't seen this field offset yet.
+        sortedFieldOrder[helperPtr->fieldOffsets[i]] = i;
+    }
+
+    // Set the layoutSizes (includes holes from alignment of the fields.)
+    int lastField = -1;
+    for (unsigned i = 0; i < SYSTEMV_MAX_NUM_FIELDS_IN_REGISTER_PASSED_STRUCT; i++)
+    {
+        int ordinal = sortedFieldOrder[i];
+        if (ordinal == -1)
+        {
+            continue;
+        }
+
+        if (lastField == -1)
+        {
+            lastField = ordinal;
+            continue;
+        }
+
+        helperPtr->fieldLayoutSizes[lastField] = helperPtr->fieldOffsets[ordinal] - helperPtr->fieldOffsets[lastField];
+
+        lastField = ordinal;
+    }
+    // Now the last field
+    _ASSERTE(lastField != -1); // if lastField==-1, then the struct has no fields!
+
+    // A field size cannot be bigger than the size of an eightbyte; any necessary padding
+    // will be handled later.
+    _ASSERTE(helperPtr->fieldSizes[lastField] <= SYSTEMV_EIGHT_BYTE_SIZE_IN_BYTES);
+    helperPtr->fieldLayoutSizes[lastField] = min(helperPtr->structSize - helperPtr->fieldOffsets[lastField], SYSTEMV_EIGHT_BYTE_SIZE_IN_BYTES);
+
+    // Add logical fields for any padding bytes at the end of the struct.
+    unsigned int paddingOffset = helperPtr->fieldOffsets[lastField] + helperPtr->fieldLayoutSizes[lastField];
+    _ASSERTE(paddingOffset == helperPtr->structSize || lastField < SYSTEMV_MAX_NUM_FIELDS_IN_REGISTER_PASSED_STRUCT);
+
+    // Padding bytes inside of an eightbyte that contains other fields must be classified as NO_CLASS
+    // as per the SysV ABI spec. Subsequent bytes will be classified as INTEGERs to ensure that they
+    // do not get dropped.
+    //
+    // NOTE: if INTEGER-classified bytes are actually shadowing SSE-classified native fields, these
+    // classifications will be incorrect. Unfortunately, there's no accurate way to guess whether or
+    // not this shadowing is occurring based on the size of the struct and the managed fields alone.
+    unsigned int nextEightByteOffset = paddingOffset + (paddingOffset % 8);
+    for (unsigned int offset = paddingOffset, ordinal = lastField + 1; offset < helperPtr->structSize; offset++, ordinal++)
+    {
+        _ASSERTE(sortedFieldOrder[offset] == -1);
+        sortedFieldOrder[offset] = ordinal;
+        helperPtr->fieldOffsets[ordinal] = offset;
+        helperPtr->fieldLayoutSizes[ordinal] = 1;
+
+        helperPtr->fieldClassifications[ordinal] = offset < nextEightByteOffset ? SystemVClassificationTypeNoClass
+                                                                                : SystemVClassificationTypeInteger;
+    }
+
```

Comment:
As we discussed, the following case needs to be handled as well.
A struct with the following layout:
4 byte int, 1 byte bool, 7 bytes hole, 1 byte bool, 1 byte bool, 2 bytes short.